drivers/adc-dac/ad74416h/ad74416h.c

```diff
@@ -0,0 +1,802 @@
+ *    on or directly connected to an Analog Devices Inc. component.
+ *
+ * THIS SOFTWARE IS PROVIDED BY ANALOG DEVICES "AS IS" AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, NON-INFRINGEMENT,
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+#include "ad74416h.h"
+#include "no_os_crc8.h"
+#include "no_os_delay.h"
+#include "no_os_error.h"
+#include "no_os_util.h"
+#include "no_os_alloc.h"
+
+/******************************************************************************/
+/********************** Macros and Constants Definitions **********************/
+/******************************************************************************/
+#define AD74416H_CRC_POLYNOMIAL 	0x7
+#define AD74416H_DIN_DEBOUNCE_LEN 	NO_OS_BIT(5)
+#define AD77416H_DEV_ADDRESS_MSK	NO_OS_GENMASK(5, 4)
+
+/******************************************************************************/
+/************************ Variable Declarations ******************************/
+/******************************************************************************/
+NO_OS_DECLARE_CRC8_TABLE(_crc_table);
+
+static const unsigned int ad74416h_debounce_map[AD74416H_DIN_DEBOUNCE_LEN] = {
+	0,     13,    18,    24,    32,    42,    56,    75,
+	100,   130,   180,   240,   320,   420,   560,   750,
+	1000,  1300,  1800,  2400,  3200,  4200,  5600,  7500,
+	10000, 13000, 18000, 24000, 32000, 42000, 56000, 75000,
+};
+
+/** The ADC conversion rates */
+static const uint32_t conv_rate_ad74416h[] = { 10, 20, 1200, 4800, 9600, 200 };
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+/**
+ * @brief Converts a millivolt value in the corresponding DAC 16 bit code.
+ * @param mvolts - The millivolts value.
+ * @param code - The resulting DAC code.
+ * @return 0 in case of success, -EINVAL otherwise
+ */
+int ad74416h_dac_voltage_to_code(uint32_t mvolts, uint32_t *code)
```

Comment:
I'd return the code and get rid of `uint32_t *code` argument.

It could be done by clipping the output to the max/min possible code instead of returning an error.

In my opinion it's best for the driver API user to not handle so many error cases, but you can leave it as-is if you prefer.